src/core/libraries/np_trophy/trophy_ui.cpp

```diff
@@ -4,71 +4,85 @@
-    newInfo.trophyName = trophyName;
-    trophyQueue.push_back(newInfo);
-}
-
 void TrophyUI::Finish() {
     RemoveLayer(this);
 }
 
-bool displayingTrophy;
-std::chrono::steady_clock::time_point trophyStartedTime;
-
 void TrophyUI::Draw() {
     const auto& io = GetIO();
 
     const ImVec2 window_size{
-        std::min(io.DisplaySize.x, 200.f),
-        std::min(io.DisplaySize.y, 75.f),
+        std::min(io.DisplaySize.x, 250.f),
+        std::min(io.DisplaySize.y, 70.f),
     };
 
-    if (trophyQueue.size() != 0) {
-        if (!displayingTrophy) {
-            displayingTrophy = true;
-            trophyStartedTime = std::chrono::steady_clock::now();
-        }
-
-        std::chrono::steady_clock::time_point timeNow = std::chrono::steady_clock::now();
-        std::chrono::seconds duration =
-            std::chrono::duration_cast<std::chrono::seconds>(timeNow - trophyStartedTime);
+    SetNextWindowSize(window_size);
+    SetNextWindowCollapsed(false);
+    SetNextWindowPos(ImVec2(io.DisplaySize.x - 250, 50));
+    KeepNavHighlight();
 
-        if (duration.count() >= 5) {
-            trophyQueue.erase(trophyQueue.begin());
-            displayingTrophy = false;
+    if (Begin("Trophy Window", nullptr,
+              ImGuiWindowFlags_NoDecoration | ImGuiWindowFlags_NoSavedSettings |
+                  ImGuiWindowFlags_NoInputs)) {
+        if (trophy_icon) {
+            Image(trophy_icon.GetTexture().im_id, ImVec2(50, 50));
+            ImGui::SameLine();
+        } else {
+            // placeholder
+            const auto pos = GetCursorScreenPos();
+            ImGui::GetWindowDrawList()->AddRectFilled(pos, pos + ImVec2{50.0f},
+                                                      GetColorU32(ImVec4{0.7f}));
+            ImGui::Indent(60);
         }
+        TextWrapped("Trophy earned!\n%s", trophy_name.c_str());
+    }
+    End();
 
-        if (trophyQueue.size() != 0) {
-            SetNextWindowSize(window_size);
-            SetNextWindowCollapsed(false);
-            SetNextWindowPos(ImVec2(io.DisplaySize.x - 200, 50));
-            KeepNavHighlight();
-
-            TrophyInfo currentTrophyInfo = trophyQueue[0];
-            if (Begin("Trophy Window", nullptr,
-                      ImGuiWindowFlags_NoDecoration | ImGuiWindowFlags_NoSavedSettings |
-                          ImGuiWindowFlags_NoInputs)) {
-                Text("Trophy earned!");
-                TextWrapped("%s", currentTrophyInfo.trophyName.c_str());
-            }
-            End();
+    trophy_timer -= io.DeltaTime;
+    if (trophy_timer <= 0) {
+        if (!trophy_queue.empty()) {
+            TrophyInfo next_trophy = trophy_queue.front();
+            trophy_queue.pop();
+            current_trophy_ui.emplace(next_trophy.trophy_icon_path, next_trophy.trophy_name);
+        } else {
+            current_trophy_ui.reset();
         }
     }
 }
+
+void AddTrophyToQueue(std::filesystem::path trophyIconPath, std::string trophyName) {
```

Comment:
Just remembered: here should have a mutex. And the pop queue too

Reply:
I'm glad you noticed this, I didn't think about a race condition at all here :P